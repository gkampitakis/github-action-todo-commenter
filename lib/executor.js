"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executor = void 0;
const github_1 = require("@actions/github");
const fs_1 = require("fs");
function executor(tags) {
    const testFolder = './';
    (0, fs_1.readdir)(testFolder, (err, files) => {
        files.forEach(file => {
            console.log(file);
        });
    });
    console.log(github_1.context.payload, tags);
}
exports.executor = executor;
