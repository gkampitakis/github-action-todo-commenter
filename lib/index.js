"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@actions/core");
const executor_1 = require("./executor");
const tags = ['TODO', 'FIXME'];
const customTags = (0, core_1.getInput)('tags').split(',');
console.log(tags, customTags);
(0, executor_1.executor)(customTags || tags);
