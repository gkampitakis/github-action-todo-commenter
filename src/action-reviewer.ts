import { info } from '@actions/core';
import { ActionReviewerOptions, Octokit } from './types';

export class ActionReviewer {
  private octokit: Octokit;
  private options: ActionReviewerOptions;

  constructor(octokit: Octokit, options: ActionReviewerOptions) {
    this.octokit = octokit;
    this.options = options;
  }

  public async createReview(body: string) {
    const { id, body: oldReviewBody } = await this.reviewExists();

    if (oldReviewBody === body) {
      info('No new changes detected since ');
      return;
    }

    if (id) {
      await this.deleteReview(id);
    }

    await this.octokit.rest.issues.createComment({
      owner: this.options.owner,
      repo: this.options.repo,
      issue_number: this.options.prNumber,
      body
    });

    return;
  }

  public async singleCommentReview(
    body: string,
    { line, path }: { line: number; path: string }
  ) {
    return this.octokit.rest.pulls.createReviewComment({
      owner: this.options.owner,
      repo: this.options.repo,
      pull_number: this.options.prNumber,
      commit_id: this.options.commitId,
      position: line,
      body,
      path,
      line
    });
  }
// test
  public async deleteReview(id: number) {
    return this.octokit.rest.issues.deleteComment({
      owner: this.options.owner,
      repo: this.options.repo,
      comment_id: id
    });
  }

  public async reviewExists(): Promise<{ id?: number; body?: string }> {
    const { data: allReviews } = await this.octokit.rest.issues.listComments({
      owner: this.options.owner,
      repo: this.options.repo,
      issue_number: this.options.prNumber
    });

    if (allReviews.length === 0) {
      return {};
    }

    const review = allReviews.find(item => {
      return (
        item.user?.login === 'github-actions[bot]' &&
        item.body?.startsWith(this.options.commentIdentifier)
      );
    });

    if (!review || !review.body) {
      return {};
    }

    // This removes the first line containing the comment identifier
    let index = review.body.match(/[\r\n]+/)?.index;
    index = index ? index + 1 : 0;

    return { id: review?.id, body: review.body.substr(index) };
  }
}
