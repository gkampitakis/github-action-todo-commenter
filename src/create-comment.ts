import { getOctokit } from '@actions/github';
type CreateCommentParams = {
  token: string;
  owner: string;
  repo: string;
  prNumber: number;
};

export async function createComment(
  params: CreateCommentParams,
  body: string,
  block: boolean
) {
  const { token, owner, repo, prNumber } = params;
  const octokit = getOctokit(token);

  const { data: allComments } = await octokit.rest.issues.listComments({
    owner,
    repo,
    issue_number: prNumber
  });

  const comment = allComments.find(item => {
    return (
      item.user?.login === 'github-actions[bot]' &&
      item.body?.startsWith('## Todo Commenter\n')
    );
  });

  if (comment) {
    return await octokit.rest.pulls.updateReviewComment({
      owner,
      repo,
      body,
      comment_id: comment.id,
      pull_number: prNumber
    });
  }

  return octokit.rest.pulls.createReview({
    owner,
    repo,
    event: block ? 'REQUEST_CHANGES' : 'COMMENT',
    pull_number: prNumber,
    body
  });
}
