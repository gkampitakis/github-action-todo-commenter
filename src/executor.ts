import { context } from '@actions/github';
import { readdir } from 'fs';

export function executor(tags: string[]) {
  const testFolder = './';

  // We need to do the checkout to that branch
  // Take the diff, read each file and find occurrences of tags

  readdir(testFolder, (err, files) => {
    files.forEach(file => {
      console.log(file);
    });
  });

  console.log(context.payload, tags);
}
