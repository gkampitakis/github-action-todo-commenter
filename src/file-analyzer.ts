import { createReadStream } from 'fs';
import { createInterface } from 'readline';
import path from 'path';

type FileAnalyzerResult = {
  file: string;
  comments: Comments;
}
type Comments = Record<string, string[]>;
type EnhancedTag = { tag: string, regex: RegExp };

export async function fileAnalyzer(filePaths: string[], tags: string[]): Promise<FileAnalyzerResult[]> {
  const enhancedTags = tags.map(tag => ({ tag, regex: new RegExp(tag + '(.*)') }));
  const ignoreMinSize = Math.min(...tags.map(t => t.length));
  const result: FileAnalyzerResult[] = [];

  for (const filePath of filePaths) {
    const absolutePath = path.join(process.cwd(), filePath);
    const comments = await readFile(absolutePath, enhancedTags, ignoreMinSize);

    result.push({
      file: filePath,
      comments
    });
  }

  return result;
}

async function readFile(filepath: string, tags: EnhancedTag[], ignoreMinSize: number): Promise<Comments> {
  const comments: Comments = {};
  const stream = createReadStream(filepath, 'utf-8');
  const rl = createInterface({
    input: stream,
    crlfDelay: Infinity
  });

  for await (const line of rl) {
    if (ignoreMinSize > line.length) continue;

    for (let { tag, regex } of tags) {
      const matched = line.match(regex);

      if (!matched) continue;

      const comment = matched[1].trim();

      if (comments[tag]) {
        comments[tag].push(comment);
        continue;
      }

      comments[tag] = [comment];
    }
  }

  return comments;
}
