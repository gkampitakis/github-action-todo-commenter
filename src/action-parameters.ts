import { getInput } from '@actions/core';
import { Context } from '@actions/github/lib/context';

import { GetActionParams, GetInputParams } from './types';

export function getInputs(): GetInputParams {
  const tags = getInput('tags') || 'TODO:,FIXME:,BUG:';
  const reviewMsg = getInput('review-message');
  const ignoreFilesPattern = getInput('ignore-pattern');
  const token = getInput('github-token', { required: true });
  const commentTitle = getInput('comment-title') || 'Todo Commenter';

  return {
    tags: tags.split(','),
    reviewMsg,
    token,
    ignoreFilesPattern,
    commentTitle
  };
}

export function getActionParameters(ctx: Context): GetActionParams {
  const {
    actor,
    repo: { owner, repo },
    eventName,
    payload: { pull_request }
  } = ctx;
  const base = pull_request?.base?.ref;
  const head = pull_request?.head?.ref;

  console.log(eventName);

  if (eventName !== 'pull_request') {
    throw new Error('Action only supports pull requests');
  }

  if (!pull_request?.number) {
    throw new Error('Action cannot identify pull request number');
  }

  return {
    actor,
    owner,
    repo,
    base,
    head,
    prNumber: pull_request?.number
  };
}
