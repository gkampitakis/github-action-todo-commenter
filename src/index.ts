import { getInput } from '@actions/core';
import { executor } from './executor';

const tags = ['TODO', 'FIXME'];
const customTags = getInput('tags').split(',');

console.log(tags, customTags);


executor(customTags || tags);
