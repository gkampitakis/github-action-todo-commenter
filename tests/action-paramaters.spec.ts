import { Context } from '@actions/github/lib/context';
import { getActionParameters, getInputs } from '../src/action-parameters';

let mockInputs: Record<string, string> = {};

jest.mock('@actions/core', () => ({
  getInput: (input: string) => mockInputs[input]
}));

describe('getActionParameters', () => {
  it('should return action parameters', () => {
    const ctx = {
      actor: 'mock-actor',
      repo: {
        owner: 'mock-owner',
        repo: 'mock-repo'
      },
      eventName: 'pull_request',
      payload: {
        pull_request: {
          number: 10,
          base: { ref: 'master' },
          head: { ref: 'pr' }
        }
      }
    } as unknown as Context;

    expect(getActionParameters(ctx)).toMatchSnapshot();
  });

  it("should throw error if event != 'pull_request'", () => {
    const ctx = {
      repo: {},
      eventName: 'push',
      payload: {}
    } as Context;

    expect(() => getActionParameters(ctx)).toThrow(
      'Action only supports pull requests'
    );
  });

  it('should throw error if cannot find pull request number', () => {
    const ctx = {
      repo: {},
      eventName: 'pull_request',
      payload: {
        pull_request: {}
      }
    } as Context;

    expect(() => getActionParameters(ctx)).toThrow(
      'Action cannot identify pull request number'
    );
  });
});

describe('getInputs', () => {
  beforeEach(() => {
    mockInputs = {};
  });

  it('should return all inputs', () => {
    mockInputs = {
      'github-token': 'mock-token',
      tags: 'test:,data:,more',
      'review-message': 'hello world',
      'ignore-pattern': 'mock-pattern',
      'comment-title': 'my title'
    };
    const inputs = getInputs();

    expect(inputs).toMatchSnapshot();
  });
});
