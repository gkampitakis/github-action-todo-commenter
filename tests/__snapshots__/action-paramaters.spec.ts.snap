// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getActionParameters should return action parameters 1`] = `
{
  "actor": "mock-actor",
  "base": "master",
  "head": "pr",
  "owner": "mock-owner",
  "prNumber": 10,
  "repo": "mock-repo",
}
`;

exports[`getInputs should return all inputs 1`] = `
{
  "commentTitle": "my title",
  "ignoreFilesPattern": "mock-pattern",
  "reviewMsg": "hello world",
  "tags": [
    "test:",
    "data:",
    "more",
  ],
  "token": "mock-token",
}
`;
